exonum/src/storage/proof_map_index/tests.rs

```diff
@@ -597,30 +1147,133 @@ fn iter(db: Box<Database>) {
     );
 }
 
-fn bytes_to_hex<T: AsRef<[u8]> + ?Sized>(bytes: &T) -> String {
-    let strings: Vec<String> = bytes
-        .as_ref()
-        .iter()
-        .map(|b| format!("{:02x}", b))
-        .collect();
-    strings.join("")
+fn tree_with_hashed_key(db: Box<Database>) {
+    use std::iter::FromIterator;
+
+    encoding_struct! {
+        struct Point {
+            x: u16,
+            y: u16,
+        }
+    }
+
+    impl HashedKey for Point {}
+
+    fn hash_isolated_node(key: &ProofPath, h: &Hash) -> Hash {
+        HashStream::new()
+            .update(&key.as_bytes())
+            .update(h.as_ref())
+            .hash()
+    }
+
+    let mut storage = db.fork();
+    let mut table = ProofMapIndex::new(IDX_NAME, &mut storage);
+
+    table.put(&Point::new(1, 2), vec![1, 2, 3]);
+    table.put(&Point::new(3, 4), vec![2, 3, 4]);
+
+    assert_eq!(table.get(&Point::new(1, 2)), Some(vec![1, 2, 3]));
+    assert_eq!(table.get(&Point::new(2, 2)), None);
+
+    let keys: HashSet<_> = table.keys().collect();
+    assert_eq!(
+        keys,
+        HashSet::from_iter(vec![Point::new(3, 4).hash(), Point::new(1, 2).hash()])
+    );
+
+    let kvs: HashSet<_> = table.iter().collect();
+    assert_eq!(
+        kvs,
+        HashSet::from_iter(vec![
+            (Point::new(3, 4).hash(), vec![2, 3, 4]),
+            (Point::new(1, 2).hash(), vec![1, 2, 3]),
+        ])
+    );
+
+    let proof = table.get_proof(Point::new(1, 2));
+    assert_eq!(
+        proof.proof_unchecked(),
+        vec![(ProofPath::new(&Point::new(3, 4)), hash(&vec![2, 3, 4]))]
+    );
+    let proof = proof.check().unwrap();
+    assert_eq!(proof.all_entries(), vec![(&Point::new(1, 2), Some(&vec![1, 2, 3]))]);
+    assert_eq!(proof.merkle_root(), table.merkle_root());
+
+    let key = Point::new(3, 4);
+    let other_key = Point::new(1, 2);
+    table.remove(&key);
+    let keys: Vec<_> = table.keys().collect();
+    assert_eq!(keys, vec![other_key.hash()]);
+    assert_eq!(table.get(&key), None);
+    assert_eq!(table.get(&other_key), Some(vec![1, 2, 3]));
+    assert_eq!(
+        table.merkle_root(),
+        hash_isolated_node(&ProofPath::new(&other_key.hash()), &hash(&vec![1, 2, 3]))
+    );
+}
+
+// // // // Macros // // // //
```

Comment:
Personally, I'm against such "fancy" comments. :upside_down_face: Probably, except for the tutorials.